pkg/controller/dnszone/dnszone_controller.go

```diff
@@ -480,3 +481,44 @@ func lookupSOARecord(zone string, logger log.FieldLogger) (bool, error) {
 	}
 	return false, nil
 }
+
+// IsErrorUpdateEvent returns true when the update event is from
+// error state.
+func IsErrorUpdateEvent(evt event.UpdateEvent) bool {
+	new, ok := evt.ObjectNew.(*hivev1.DNSZone)
+	if !ok {
+		return false
+	}
+	if len(new.Status.Conditions) == 0 {
+		return false
+	}
+
+	old, ok := evt.ObjectOld.(*hivev1.DNSZone)
+	if !ok {
+		return false
+	}
+
+	errorConds := []hivev1.DNSZoneConditionType{
+		hivev1.InsufficientCredentialsCondition,
+		hivev1.AuthenticationFailureCondition,
+	}
+
+	for _, cond := range errorConds {
+		cn := controllerutils.FindDNSZoneCondition(new.Status.Conditions, cond)
+		if cn != nil && cn.Status == corev1.ConditionTrue {
+			co := controllerutils.FindDNSZoneCondition(old.Status.Conditions, cond)
+			if co == nil {
+				return true // newly added failure condition
+			}
+			if co.Status != corev1.ConditionTrue {
+				return true // newly Failed failure condition
+			}
+			if cn.Message != co.Message ||
+				cn.Reason != co.Reason {
+				return true // already failing but change in error reported
+			}
+		}
+	}
+
+	return false
+}
```

Comment:
This is great. Feels like there's potential here where we could just pass our error conditions to controllerutils.NewRateLimitedUpdateEventHandler maybe in conjunction with https://issues.redhat.com/browse/HIVE-1521 in the future.